File: s3/credentials.go
package s3

import (
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
)

const (
	MinLeg_AccessKey  = 6
	MaxLeg_AccessKey  = 20
	MinLeg_SecretKey  = 6
	MaxLeg_SecretKey  = 40
	alphaNumericTable = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	// specialCharTable  = "!@#$%^&*()-_+={}[]|:;<>,.?/"
	reservedChars = "=,"
)

var (
	ErrInvalidAccessKeyLength   = fmt.Errorf("access key length should be between %d and %d", MinLeg_AccessKey, MaxLeg_AccessKey)
	ErrInvalidSecretKeyLength   = fmt.Errorf("secret key length should be between %d and %d", MinLeg_SecretKey, MaxLeg_SecretKey)
	ErrNoAccessKeyWithSecretKey = fmt.Errorf("access key must be specified if secret key is specified")
	ErrNoSecretKeyWithAccessKey = fmt.Errorf("secret key must be specified if access key is specified")
	ErrContainsReservedChars    = fmt.Errorf("access key contains one of reserved characters '=' or ','")
)

// Default credentials (if youre too stupid to setup your own)
const (
	DefaultAccessKey = "lumiserver"
	DefaultSecretKey = "lumiserver"
)

func ContainsReservedChars(s string) bool {
	return strings.ContainsAny(s, reservedChars)
}

type Credentials struct {
	AccessKey string
	SecretKey string
}

func GenerateAccessKey(length int, random io.Reader) (string, error) {
	if random == nil {
		random = rand.Reader
	}
	if length <= 0 {
		length = MaxLeg_AccessKey
	}
	if length < MinLeg_AccessKey {
		return "", errors.New("auth: access key length is too short")
	}

	key := make([]byte, length)
	if _, err := io.ReadFull(random, key); err != nil {
		return "", err
	}
	for i := range key {
		key[i] = alphaNumericTable[key[i]%byte(len(alphaNumericTable))]
	}
	return string(key), nil
}

func GenerateSecretKey(length int, random io.Reader) (string, error) {
	if random == nil {
		random = rand.Reader
	}
	if length <= 0 {
		length = MaxLeg_SecretKey
	}
	if length < MinLeg_SecretKey {
		return "", errors.New("auth: secret key length is too short")
	}

	key := make([]byte, base64.RawStdEncoding.DecodedLen(length))
	if _, err := io.ReadFull(random, key); err != nil {
		return "", err
	}

	s := base64.RawStdEncoding.EncodeToString(key)
	return strings.ReplaceAll(s, "/", "+"), nil
}

func ValidateAccessKey(r *http.Request, accessKey string) (Credentials, bool, ErrorCode) {
	cred, isHeader, err := getRequestKey(r)
	if err != ErrNone {
		return Credentials{}, false, err
	}

	if cred.AccessKey == "" {
		return Credentials{}, false, ErrMissingCredTag
	}
	if accessKey != "" && cred.AccessKey != accessKey {
		return Credentials{}, false, ErrInvalidAccessKeyID
	}
	return cred, isHeader, ErrNone

}

File: crates/credentials/src/credentials.rs
use crate::{
    ALPHA_NUMERIC_TABLE, 
    MAX_LEG_ACCESS_KEY, 
    MAX_LEG_SECRET_KEY, 
    MIN_LEG_ACCESS_KEY, 
    MIN_LEG_SECRET_KEY
};

use rand::Rng;

#[derive(Debug, Clone, PartialEq)]
pub enum KeyError {
   InvalidAccessKeyLength,
   InvalidSecretKeyLength,
   NoAccessKeyWithSecretKey,
   NoSecretKeyWithAccessKey,
   ContainsReservedChars,
   AccessKeyTooShort,
   SecretKeyTooShort,
   IoError(String),
}

impl std::fmt::Display for KeyError {
   fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
       match self {
           KeyError::InvalidAccessKeyLength => write!(f, "access key length should be between {} and {}", MIN_LEG_ACCESS_KEY, MAX_LEG_ACCESS_KEY),
           KeyError::InvalidSecretKeyLength => write!(f, "secret key length should be between {} and {}", MIN_LEG_SECRET_KEY, MAX_LEG_SECRET_KEY),
           KeyError::NoAccessKeyWithSecretKey => write!(f, "access key must be specified if secret key is specified"),
           KeyError::NoSecretKeyWithAccessKey => write!(f, "secret key must be specified if access key is specified"),
           KeyError::ContainsReservedChars => write!(f, "access key contains one of reserved characters '=' or ','"),
           KeyError::AccessKeyTooShort => write!(f, "auth: access key length is too short"),
           KeyError::SecretKeyTooShort => write!(f, "auth: secret key length is too short"),
           KeyError::IoError(msg) => write!(f, "IO error: {}", msg),
       }
   }
}

pub fn generate_access_key(length: usize) -> Result<String, KeyError> {
        if length < MIN_LEG_ACCESS_KEY {
            return Err(KeyError::AccessKeyTooShort)
        }
        if length > MAX_LEG_ACCESS_KEY {
            return Err(KeyError::InvalidAccessKeyLength)
        }
        let mut result = String::with_capacity(length);
        let mut rng = rand::rng();
        for _ in 0..length {
            let idx = rng.random_range(0..ALPHA_NUMERIC_TABLE.len());
            result.push(ALPHA_NUMERIC_TABLE[idx] as char);
        }
        Ok(result)
}

// same shit pretty much :/
pub fn generate_secret_key(length: usize) -> Result<String, KeyError> {
        if length < MIN_LEG_SECRET_KEY {
            return Err(KeyError::SecretKeyTooShort)
        }
        if length > MAX_LEG_SECRET_KEY {
            return Err(KeyError::InvalidSecretKeyLength)
        }
        let mut result = String::with_capacity(length);
        let mut rng = rand::rng();
        for _ in 0..length {
            let idx = rng.random_range(0..ALPHA_NUMERIC_TABLE.len());
            result.push(ALPHA_NUMERIC_TABLE[idx] as char);
        }
        Ok(result)
}


#[derive(Debug, Clone, PartialEq)]
pub struct Credentials {
   pub access_key: String,
   pub secret_key: String,
}